visualizer.py

```diff
@@ -0,0 +1,323 @@
+
+            self.traces[trace.label].values[self.trace_index:self.trace_index+num_elements] = trace.values[:num_elements]
+
+        # Extend the timebase
+        sample_index = frame.header.sequence * frame.header.batch_size
+        self.times[self.trace_index:self.trace_index+num_elements] = \
+            [np.uint32(sample_index + offset & 0xFFFFFFFF) for offset in range(num_elements)]
+
+        self.trace_index += num_elements
+
+
+    async def _update(self):
+        """ Ingest a single livestream frame. """
+        # Ingest stream frames.
+        frame = await self._stream.queue.get()
+
+        if self.trigger is not TriggerState.TRIGGERED:
+            return
+
+        if frame.format_id != self._current_format:
+            self.traces = dict()
+            self.times = np.empty(self.max_size, dtype=np.uint32)
+            self._current_format = frame.format_id
+
+        # Append trace data
+        self._append_traces(frame)
+
+        # Check for the capture buffer being filled.
+        if self.trace_index >= self.max_size:
+            logging.info('Got %d samples - stopping trigger', self.max_size)
+            self.trigger = TriggerState.STOPPED
+
+
+    def get_traces(self):
+        """ Get the traces and times to display in the graph. """
+        return (self.times - self.times[0]), self.traces
+
+
+class StreamVisualizer:
+    """ A visualizer for stabilizer's livestream data. """
+
+    def __init__(self, receiver: StreamReceiver, document: bokeh.document.Document):
+        """ Initialize the visualizer.
+
+        Args:
+            receiver: The StreamReceiver object for managing stream reception.
+            document: The Bokeh document to draw into.
+        """
+
+        self.receiver = receiver
+
+        figure = bokeh.plotting.figure(output_backend="webgl", sizing_mode='stretch_both')
+
+        # Add a trigger button
+        trigger_button = bokeh.models.Button(label='Single', button_type='primary')
+        trigger_button.on_click(self.capture)
+
+        force_button = bokeh.models.Button(label='Force', button_type='primary')
+        force_button.on_click(self.receiver.force_trigger)
+
+        self._capture_duration_input = bokeh.models.TextInput(title='Capture Duration',
+                                                              value='0.001',
+                                                              width=100,
+                                                              sizing_mode='fixed')
+        self._capture_duration_input.on_change('value', self.handle_duration)
+        self._trigger_state = bokeh.models.Div(text='Trigger State: <b>IDLE</b>')
+
+        control_layout = bokeh.layouts.column(self._trigger_state, trigger_button, force_button,
+                                              self._capture_duration_input)
+
+        self.layout = bokeh.layouts.row(figure, control_layout, sizing_mode='stretch_height')
```

Comment:
Do we need to change the defaults (sizing_mode) here? Let's start with a small plot.

Reply:
I believe the plot by-default is quite small using fixed sizing. I was aiming for the plot to be a majority of the screen - is there a desire for it to be fixed-size and/or smaller? Or is that largely in hopes of increasing rendering speeds?